python/ray/tune/integration/docker.py

```diff
@@ -0,0 +1,177 @@
+                 sync_config=tune.SyncConfig(
+                     sync_to_driver=DockerSyncer))
+
+    """
+
+    _cluster_path = os.path.expanduser("~/ray_bootstrap_config.yaml")
+
+    def __init__(self, local_dir, remote_dir, sync_client=None):
+        self.local_ip = services.get_node_ip_address()
+        self.worker_ip = None
+        self.worker_node_id = None
+
+        global _provider
+        if _provider is None:
+            _provider = _WrappedProvider(self._cluster_path)
+
+        sync_client = sync_client or DockerSyncClient()
+        sync_client.configure(_provider, _provider.config)
+
+        self._provider = _provider
+
+        super(NodeSyncer, self).__init__(local_dir, remote_dir, sync_client)
+
+    def set_worker_ip(self, worker_ip):
+        self.worker_ip = worker_ip
+        self.worker_node_id = self._provider.get_ip_to_node_id()[1][worker_ip]
+
+    @property
+    def _remote_path(self):
+        return (self.worker_node_id, self._remote_dir)
+
+
+class DockerSyncClient(SyncClient):
+    """DockerSyncClient to be used by DockerSyncer.
+    This client takes care of executing the synchronization
+    commands for Docker nodes. In its ``sync_down`` and
+    ``sync_up`` commands, it expects tuples for the source
+    and target, respectively, for compatibility with the
+    DockerCommandRunner.
+    """
+
+    def __init__(self):
+        self._command_runners = {}
+        self.provider = None
+
+    def configure(self, provider, cluster_config):
+        self.provider = provider
+        self.cluster_config = cluster_config
+
+    def _create_command_runner(self, node_id):
+        """Create a command runner for one Docker node"""
+        args = {
+            "log_prefix": "DockerSync: ",
+            "node_id": node_id,
+            "provider": self.provider,
+            "auth_config": self.cluster_config["auth"],
+            "cluster_name": self.cluster_config["cluster_name"],
+            "process_runner": subprocess,
+            "use_internal_ip": True,
+            "docker_config": self.cluster_config["docker"],
+        }
+        return DockerCommandRunner(**args)
+
+    def _get_command_runner(self, node_id):
+        """Create command runner if it doesn't exist"""
+        # Todo(krfricke): These cached runners are currently
+        # never cleaned up. They are cheap so this shouldn't
+        # cause much problems, but should be addressed if
+        # the SyncClient is used more extensively in the future.
+        if node_id not in self._command_runners:
+            command_runner = self._create_command_runner(node_id)
+            self._command_runners[node_id] = command_runner
+        return self._command_runners[node_id]
+
+    def sync_up(self, source, target):
+        """Here target is a tuple (target_node, target_dir)"""
+        target_node, target_dir = target
+
+        # Add trailing slashes for rsync
+        source += "/" if not source.endswith("/") else ""
```

Comment:
```suggestion
        source = os.path.join(source, "")
```

nit: could be shorter :)

Reply:
Always more to learn!